server/etcdserver/txn/txn.go

```diff
@@ -43,16 +41,23 @@ func Put(ctx context.Context, lg *zap.Logger, lessor lease.Lessor, kv mvcc.KV, t
 			traceutil.Field{Key: "req_size", Value: p.Size()},
 		)
 	}
-	val, leaseID := p.Value, lease.LeaseID(p.Lease)
-	if txnWrite == nil {
-		if leaseID != lease.NoLease {
-			if l := lessor.Lookup(leaseID); l == nil {
-				return nil, nil, lease.ErrLeaseNotFound
-			}
+	leaseID := lease.LeaseID(p.Lease)
+	if leaseID != lease.NoLease {
+		if l := lessor.Lookup(leaseID); l == nil {
+			return nil, nil, lease.ErrLeaseNotFound
```

Comment:
Only perform this check in Put? Why not in `put`? Note that you will miss the check when put is called inside a txn.

Suggestion: move the check into function `put`

Reply:
Please notice that I `put` function is only called via `applyTxn` and provided `writeTxn` is never nil. There is no need for `put` for checking lessor after separation from `Put`. 

To explain why it's done, please read the `Txn` code. Txn is executed in two phases, first with readTxn and second with writeTxn. It does all validation including checking the lease in the first phase, and only calls calls function `put` in second phase. It means that there is no need for `put` to validate the lease again.